Multiple file upload using Quarkus and RestEasy reactive

Trying to upload multiple files along with file details . want to map the request to POJO

```java
public class MulitipleDocumentDetailsRequest implements Serializable {

	private static final long serialVersionUID = 1L;

	@RestForm("fileUpload")
	private List<FileUpload> fileUpload;

	@RestForm("fileDetails")
	@PartType(MediaType.APPLICATION_JSON)
	private List<FileDetails> fileDetails;

}
```

no of files uploading to the system is dynamic hence defined FileUpload as list.

endpoint:
	
```java
       @POST
	@Path("/v1/checkinMultipleDoc")
	@Consumes(MediaType.MULTIPART_FORM_DATA)
	@Produces(MediaType.TEXT_PLAIN)
	public Response checkinMultipleDoc(@MultipartForm MulitipleDocumentDetailsRequest documentDetailsRequests){

       }
```

Exception:

```
2021-05-05 23:25:12,540 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (executor-thread-1) Request failed: javax.ws.rs.WebApplicationException: HTTP 400 Bad Request
        at org.jboss.resteasy.reactive.server.handlers.ParameterHandler.handle(ParameterHandler.java:58)
        at org.jboss.resteasy.reactive.server.handlers.ParameterHandler.handle(ParameterHandler.java:14)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:122)
        at io.quarkus.runtime.CleanableExecutor$CleaningRunnable.run(CleanableExecutor.java:231)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2415)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at java.lang.Thread.run(Thread.java:748)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 0
        at org.jboss.resteasy.reactive.server.handlers.ParameterHandler.handleResult(ParameterHandler.java:116)
        at org.jboss.resteasy.reactive.server.handlers.ParameterHandler.handle(ParameterHandler.java:52)
        ... 11 more
```

Details :
Java version : 1.8 
Quarkus version 1.13.0.Final

--- a/docs/src/main/asciidoc/resteasy-reactive.adoc
+++ b/docs/src/main/asciidoc/resteasy-reactive.adoc
@@ -426,6 +426,8 @@ while the `file` field will contain data about the uploaded file in the `image` 
 
 NOTE: `FileUpload` provides access to various metadata of the uploaded file. If however all you need is a handle to the uploaded file, `java.nio.file.Path` or `java.io.File` could be used.
 
+NOTE: When access to all uploaded files without specifying the form names is needed, RESTEasy Reactive allows the use of `@RestForm List<FileUpload>`, where it is important to **not** set a name for the `@RestForm` annotation.
+
 NOTE: `@PartType` is used to aid in deserialization of the corresponding part of the request into the desired Java type. It is very useful when for example the corresponding body part is JSON
 and needs to be converted to a POJO.
 

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/MultipartPopulatorGenerator.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/MultipartPopulatorGenerator.java
@@ -16,6 +16,7 @@ import org.jboss.jandex.DotName;
 import org.jboss.jandex.FieldInfo;
 import org.jboss.jandex.IndexView;
 import org.jboss.jandex.MethodInfo;
+import org.jboss.jandex.ParameterizedType;
 import org.jboss.jandex.Type;
 import org.jboss.logging.Logger;
 import org.jboss.resteasy.reactive.common.processor.AsmUtil;
@@ -262,9 +263,7 @@ final class MultipartPopulatorGenerator {
                         }
                     }
 
-                    boolean isFileRelatedField = fieldDotName.equals(DotNames.FIELD_UPLOAD_NAME)
-                            || fieldDotName.equals(DotNames.PATH_NAME)
-                            || fieldDotName.equals(DotNames.FILE_NAME);
+                    boolean isFileRelatedField = isFileRelatedType(fieldDotName);
                     ResultHandle formAttrNameHandle = populate.load(formAttrName);
                     AssignableResultHandle resultHandle = populate.createVariable(Object.class);
 
@@ -295,6 +294,9 @@ final class MultipartPopulatorGenerator {
                             fileUploadFalse.breakScope();
                         }
                     } else {
+                        // this is a common enough mistake, so let's provide a good error message
+                        failIfFileTypeUsedAsGenericType(field, fieldType, fieldDotName);
+
                         if (fieldDotName.equals(DotNames.STRING_NAME) && partType.equals(MediaType.TEXT_PLAIN)) {
                             // in this case all we need to do is read the value of the form attribute
 
@@ -388,5 +390,24 @@ final class MultipartPopulatorGenerator {
             populate.returnValue(null);
         }
         return generateClassName;
+    }
+
+    private static boolean isFileRelatedType(DotName type) {
+        return type.equals(DotNames.FIELD_UPLOAD_NAME) || type.equals(DotNames.PATH_NAME) || type.equals(DotNames.FILE_NAME);
+    }
+
+    private static void failIfFileTypeUsedAsGenericType(FieldInfo field, Type fieldType, DotName fieldDotName) {
+        if (fieldType.kind() == Type.Kind.PARAMETERIZED_TYPE) {
+            ParameterizedType parameterizedType = fieldType.asParameterizedType();
+            if (!parameterizedType.arguments().isEmpty()) {
+                DotName argTypeDotName = parameterizedType.arguments().get(0).name();
+                if (isFileRelatedType(argTypeDotName)) {
+                    throw new IllegalArgumentException("Type '" + argTypeDotName.withoutPackagePrefix()
+                            + "' cannot be used as a generic type of '"
+                            + fieldDotName.withoutPackagePrefix() + "'. Offending field is '" + field.name() + "' of class '"
+                            + field.declaringClass().name() + "'");
+                }
+            }
+        }
     }
 }

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/MultipartPopulatorGenerator.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/MultipartPopulatorGenerator.java
@@ -20,6 +20,7 @@ import org.jboss.jandex.ParameterizedType;
 import org.jboss.jandex.Type;
 import org.jboss.logging.Logger;
 import org.jboss.resteasy.reactive.common.processor.AsmUtil;
+import org.jboss.resteasy.reactive.common.processor.ResteasyReactiveDotNames;
 import org.jboss.resteasy.reactive.common.processor.TypeArgMapper;
 import org.jboss.resteasy.reactive.common.util.DeploymentUtils;
 import org.jboss.resteasy.reactive.common.util.types.TypeSignatureParser;
@@ -248,8 +249,10 @@ final class MultipartPopulatorGenerator {
                     }
 
                     String formAttrName = field.name();
+                    boolean formAttrNameSet = false;
                     AnnotationValue formParamValue = formParamInstance.value();
                     if (formParamValue != null) {
+                        formAttrNameSet = true;
                         formAttrName = formParamValue.asString();
                     }
 
@@ -263,11 +266,10 @@ final class MultipartPopulatorGenerator {
                         }
                     }
 
-                    boolean isFileRelatedField = isFileRelatedType(fieldDotName);
                     ResultHandle formAttrNameHandle = populate.load(formAttrName);
                     AssignableResultHandle resultHandle = populate.createVariable(Object.class);
 
-                    if (isFileRelatedField) {
+                    if (isFileRelatedType(fieldDotName)) {
                         // uploaded file are present in the RoutingContext and are extracted using MultipartSupport#getFileUpload
 
                         ResultHandle fileUploadHandle = populate.invokeStaticMethod(
@@ -293,6 +295,21 @@ final class MultipartPopulatorGenerator {
                             }
                             fileUploadFalse.breakScope();
                         }
+                    } else if (isListOfFileUpload(fieldType)) {
+                        // in this case we allow injection of all the uploaded file as long as a name
+                        // was not provided in @RestForm (which makes no semantic sense)
+                        if (formAttrNameSet) {
+                            throw new IllegalArgumentException(
+                                    "When using a 'List<FileUpload>' field to capture all uploaded files, " +
+                                            "the field must be annotated '@RestForm' without providing a name. " +
+                                            "Offending field is '" + field.name() + "' of class '"
+                                            + field.declaringClass().name() + "'");
+                        }
+                        ResultHandle allFileUploadsHandle = populate.invokeStaticMethod(
+                                MethodDescriptor.ofMethod(MultipartSupport.class, "getFileUploads", List.class,
+                                        ResteasyReactiveRequestContext.class),
+                                rrCtxHandle);
+                        populate.assign(resultHandle, allFileUploadsHandle);
                     } else {
                         // this is a common enough mistake, so let's provide a good error message
                         failIfFileTypeUsedAsGenericType(field, fieldType, fieldDotName);
@@ -394,6 +411,17 @@ final class MultipartPopulatorGenerator {
 
     private static boolean isFileRelatedType(DotName type) {
         return type.equals(DotNames.FIELD_UPLOAD_NAME) || type.equals(DotNames.PATH_NAME) || type.equals(DotNames.FILE_NAME);
+    }
+
+    private static boolean isListOfFileUpload(Type fieldType) {
+        if ((fieldType.kind() == Type.Kind.PARAMETERIZED_TYPE) && fieldType.name().equals(ResteasyReactiveDotNames.LIST)) {
+            ParameterizedType parameterizedType = fieldType.asParameterizedType();
+            if (!parameterizedType.arguments().isEmpty()) {
+                DotName argTypeDotName = parameterizedType.arguments().get(0).name();
+                return argTypeDotName.equals(DotNames.FIELD_UPLOAD_NAME);
+            }
+        }
+        return false;
     }
 
     private static void failIfFileTypeUsedAsGenericType(FieldInfo field, Type fieldType, DotName fieldDotName) {

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/FormDataWithAllUploads.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/FormDataWithAllUploads.java
@@ -1,0 +1,47 @@
+package io.quarkus.resteasy.reactive.server.test.multipart;
+
+import java.util.List;
+
+import javax.ws.rs.core.MediaType;
+
+import org.jboss.resteasy.reactive.PartType;
+import org.jboss.resteasy.reactive.RestForm;
+import org.jboss.resteasy.reactive.multipart.FileUpload;
+
+public class FormDataWithAllUploads extends FormDataBase {
+
+    @RestForm
+    // don't set a part type, use the default
+    private String name;
+
+    @RestForm
+    @PartType(MediaType.TEXT_PLAIN)
+    private Status status;
+
+    @RestForm
+    private List<FileUpload> uploads;
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    public Status getStatus() {
+        return status;
+    }
+
+    public void setStatus(Status status) {
+        this.status = status;
+    }
+
+    public List<FileUpload> getUploads() {
+        return uploads;
+    }
+
+    public void setUploads(List<FileUpload> uploads) {
+        this.uploads = uploads;
+    }
+}

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartInputWithAllUploadsTest.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartInputWithAllUploadsTest.java
@@ -1,0 +1,79 @@
+package io.quarkus.resteasy.reactive.server.test.multipart;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.function.Supplier;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.resteasy.reactive.server.test.multipart.other.OtherPackageFormDataBase;
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+
+public class MultipartInputWithAllUploadsTest extends AbstractMultipartTest {
+
+    private static final Path uploadDir = Paths.get("file-uploads");
+
+    @RegisterExtension
+    static QuarkusUnitTest test = new QuarkusUnitTest()
+            .setArchiveProducer(new Supplier<JavaArchive>() {
+                @Override
+                public JavaArchive get() {
+                    return ShrinkWrap.create(JavaArchive.class)
+                            .addClasses(FormDataBase.class, OtherPackageFormDataBase.class, FormDataWithAllUploads.class,
+                                    Status.class, MultipartResourceWithAllUploads.class)
+                            .addAsResource(new StringAsset(
+                                    // keep the files around so we can assert the outcome
+                                    "quarkus.http.body.delete-uploaded-files-on-end=false\nquarkus.http.body.uploads-directory="
+                                            + uploadDir.toString() + "\n"),
+                                    "application.properties");
+                }
+
+            });
+
+    private final File HTML_FILE = new File("./src/test/resources/test.html");
+    private final File XML_FILE = new File("./src/test/resources/test.html");
+    private final File TXT_FILE = new File("./src/test/resources/lorem.txt");
+
+    @BeforeEach
+    public void assertEmptyUploads() {
+        Assertions.assertTrue(isDirectoryEmpty(uploadDir));
+    }
+
+    @AfterEach
+    public void clearDirectory() {
+        clearDirectory(uploadDir);
+    }
+
+    @Test
+    public void testSimple() throws IOException {
+        RestAssured.given()
+                .multiPart("name", "Alice")
+                .multiPart("active", "true")
+                .multiPart("num", "25")
+                .multiPart("status", "WORKING")
+                .multiPart("htmlFile", HTML_FILE, "text/html")
+                .multiPart("xmlFile", XML_FILE, "text/xml")
+                .multiPart("txtFile", TXT_FILE, "text/plain")
+                .accept("text/plain")
+                .when()
+                .post("/multipart-all/simple/2")
+                .then()
+                .statusCode(200)
+                .body(equalTo("Alice - true - 50 - WORKING - 3 - text/plain"));
+
+        // ensure that the 3 uploaded files where created on disk
+        Assertions.assertEquals(3, uploadDir.toFile().listFiles().length);
+    }
+}

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartResourceWithAllUploads.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartResourceWithAllUploads.java
@@ -1,0 +1,30 @@
+package io.quarkus.resteasy.reactive.server.test.multipart;
+
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+import org.jboss.resteasy.reactive.MultipartForm;
+import org.jboss.resteasy.reactive.multipart.FileUpload;
+
+import io.quarkus.runtime.BlockingOperationControl;
+
+@Path("/multipart-all")
+public class MultipartResourceWithAllUploads {
+
+    @POST
+    @Produces(MediaType.TEXT_PLAIN)
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    @Path("/simple/{times}")
+    public String simple(@MultipartForm FormDataWithAllUploads formData, Integer times) {
+        if (BlockingOperationControl.isBlockingAllowed()) {
+            throw new RuntimeException("should not have dispatched");
+        }
+        FileUpload txtFile = formData.getUploads().stream().filter(f -> f.name().equals("txtFile")).findFirst().get();
+        return formData.getName() + " - " + formData.active + " - " + times * formData.getNum() + " - " + formData.getStatus()
+                + " - " + formData.getUploads().size() + " - " + txtFile.contentType();
+    }
+
+}

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartSupport.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartSupport.java
@@ -6,6 +6,7 @@ import java.io.InputStream;
 import java.lang.annotation.Annotation;
 import java.lang.reflect.Type;
 import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Set;
 
@@ -93,6 +94,16 @@ public final class MultipartSupport {
             }
         }
         return null;
+    }
+
+    public static List<QuarkusFileUpload> getFileUploads(ResteasyReactiveRequestContext context) {
+        RoutingContext routingContext = context.unwrap(RoutingContext.class);
+        Set<FileUpload> fileUploads = routingContext.fileUploads();
+        List<QuarkusFileUpload> result = new ArrayList<>(fileUploads.size());
+        for (FileUpload fileUpload : fileUploads) {
+            result.add(new QuarkusFileUpload(fileUpload));
+        }
+        return result;
     }
 
     private static ByteArrayInputStream formAttributeValueToInputStream(String formAttributeValue) {

